File: apps/host/internal/githubhost/mapping.go
package githubhost

import (
	"context"
	"errors"

	pb "armadra.local/host/gen/armadra/v1"
	"armadra.local/host/internal/githubapi"
	"armadra.local/host/internal/storage"
	"google.golang.org/protobuf/proto"
)

// Status mapping: one repository, one primary source, and the groups a panel
// shows Issues under (design §7.2).
//
// Both coupling directions are configurable — a group may set the Issue state
// when entered, and an observed state may select a group — which together can
// describe a loop. A loop would make a move oscillate against the remote
// forever, so ValidateMapping refuses it before anything is stored.

// ValidateMapping checks one configuration completely. Every refusal is a
// stable reason code, because these are configuration mistakes a user has to
// find and fix, not transport failures.
func ValidateMapping(mapping *pb.GithubStatusMapping) (string, error) {
	if mapping == nil {
		return "MAPPING_REQUIRED", ErrInvalid
	}
	if len(mapping.Groups) > MaxGroups {
		return "TOO_MANY_GROUPS", ErrInvalid
	}
	switch mapping.Source {
	case pb.GithubStatusSource_GITHUB_STATUS_SOURCE_NONE:
		if len(mapping.Groups) > 0 || len(mapping.StateGroups) > 0 {
			return "SOURCE_NONE_HAS_GROUPS", ErrInvalid
		}
		if mapping.ProjectId != "" || mapping.ProjectFieldId != "" {
			return "SOURCE_NONE_HAS_PROJECT", ErrInvalid
		}
		return "", nil
	case pb.GithubStatusSource_GITHUB_STATUS_SOURCE_LABEL:
		if mapping.ProjectId != "" || mapping.ProjectFieldId != "" {
			// A repository has exactly one primary source; carrying project
			// identifiers on a label mapping would make that ambiguous.
			return "LABEL_SOURCE_HAS_PROJECT", ErrInvalid
		}
	case pb.GithubStatusSource_GITHUB_STATUS_SOURCE_PROJECT_FIELD:
		if mapping.ProjectId == "" || mapping.ProjectFieldId == "" {
			return "PROJECT_IDS_REQUIRED", ErrInvalid
		}
	default:
		return "SOURCE_UNSPECIFIED", ErrInvalid
	}
	if len(mapping.Groups) == 0 {
		return "GROUPS_REQUIRED", ErrInvalid
	}
	ids := map[string]bool{}
	labels := map[string]bool{}
	options := map[string]bool{}
	for _, group := range mapping.Groups {
		if !groupPattern.MatchString(group.Id) || ids[group.Id] {
			return "GROUP_ID_INVALID", ErrInvalid
		}
		ids[group.Id] = true
		if group.Title == "" || len(group.Title) > 128 {
			return "GROUP_TITLE_INVALID", ErrInvalid
		}
		switch group.CouplesIssueState {
		case pb.GithubIssueState_GITHUB_ISSUE_STATE_UNSPECIFIED,
			pb.GithubIssueState_GITHUB_ISSUE_STATE_OPEN,
			pb.GithubIssueState_GITHUB_ISSUE_STATE_CLOSED:
		default:
			return "GROUP_STATE_INVALID", ErrInvalid
		}
		if mapping.Source == pb.GithubStatusSource_GITHUB_STATUS_SOURCE_LABEL {
			if group.Label == "" || len(group.Label) > 128 || group.ProjectOptionId != "" {
				return "GROUP_LABEL_INVALID", ErrInvalid
			}
			// Two groups claiming one label would make an Issue's group
			// ambiguous every time, not just on conflict.
			if labels[group.Label] {
				return "GROUP_LABEL_DUPLICATE", ErrInvalid
			}
			labels[group.Label] = true
		} else {
			if group.ProjectOptionId == "" || len(group.ProjectOptionId) > 256 {
				return "GROUP_OPTION_INVALID", ErrInvalid
			}
			if options[group.ProjectOptionId] {
				return "GROUP_OPTION_DUPLICATE", ErrInvalid
			}
			options[group.ProjectOptionId] = true
		}
	}
	states := map[pb.GithubIssueState]string{}
	for _, coupling := range mapping.StateGroups {
		if coupling.State != pb.GithubIssueState_GITHUB_ISSUE_STATE_OPEN && coupling.State != pb.GithubIssueState_GITHUB_ISSUE_STATE_CLOSED {
			return "COUPLING_STATE_INVALID", ErrInvalid
		}
		if !ids[coupling.GroupId] {
			return "COUPLING_GROUP_UNKNOWN", ErrInvalid
		}
		if _, seen := states[coupling.State]; seen {
			return "COUPLING_STATE_DUPLICATE", ErrInvalid
		}
		states[coupling.State] = coupling.GroupId
	}
	if reason := detectCycle(mapping.Groups, states); reason != "" {
		return reason, ErrInvalid
	}
	return "", nil
}

// detectCycle walks group -> coupled Issue state -> that state's group. A group
// that maps back to itself is a fixed point and is fine; any longer cycle means
// entering one group would move the Issue to another and back, so it is
// refused.
func detectCycle(groups []*pb.GithubStatusGroup, states map[pb.GithubIssueState]string) string {
	next := map[string]string{}
	for _, group := range groups {
		if group.CouplesIssueState == pb.GithubIssueState_GITHUB_ISSUE_STATE_UNSPECIFIED {
			continue
		}
		target, ok := states[group.CouplesIssueState]
		if !ok || target == group.Id {
			continue
		}
		next[group.Id] = target
	}
	const (
		unvisited = 0
		onPath    = 1
		settled   = 2
	)
	mark := map[string]int{}
	for _, group := range groups {
		if mark[group.Id] != unvisited {
			continue
		}
		node := group.Id
		path := []string{}
		for {
			if mark[node] == onPath {
				return "MAPPING_CYCLE"
			}
			if mark[node] == settled {
				break
			}
			mark[node] = onPath
			path = append(path, node)
			target, ok := next[node]
			if !ok {
				break
			}
			node = target
		}
		for _, visited := range path {
			mark[visited] = settled
		}
	}
	return ""
}

// storedMapping reads one repository's configuration. A repository with none
// yet returns an empty NONE mapping at revision zero rather than an error: not
// configured is a normal state the panel shows.
func (s *Service) storedMapping(ctx context.Context, workspace string, ref *pb.GithubRepositoryRef) (*pb.GithubStatusMapping, error) {
	record, err := s.store.GithubStatusMapping(ctx, workspace, key(ref))
	if errors.Is(err, storage.ErrNotFound) {
		return &pb.GithubStatusMapping{Repository: ref, Source: pb.GithubStatusSource_GITHUB_STATUS_SOURCE_NONE}, nil
	}
	if err != nil {
		return nil, err
	}
	mapping := new(pb.GithubStatusMapping)
	if err = proto.Unmarshal(record.Mapping, mapping); err != nil {
		return nil, storage.ErrCorrupt
	}
	mapping.Repository = ref
	mapping.Revision = record.Revision
	mapping.UpdatedAtUnixMs = record.UpdatedAtMS
	return mapping, nil
}

// GetStatusMapping returns the configuration a panel groups Issues by.
func (s *Service) GetStatusMapping(ctx context.Context, caller Caller, ref *pb.GithubRepositoryRef) (*pb.GithubStatusMapping, error) {
	if err := s.authorize(caller, ScopeRead); err != nil {
		return nil, err
	}
	repository, err := s.repository(ctx, ref)
	if err != nil {
		return nil, err
	}
	return s.storedMapping(ctx, caller.WorkspaceID, repository)
}

// PutStatusMapping validates and stores one repository's configuration under
// revision CAS.
func (s *Service) PutStatusMapping(ctx context.Context, caller Caller, mapping *pb.GithubStatusMapping, expectedRevision uint64) (*pb.GithubStatusMapping, error) {
	if err := s.authorize(caller, ScopeWrite); err != nil {
		return nil, err
	}
	if mapping == nil {
		return nil, ErrInvalid
	}
	repository, err := s.repository(ctx, mapping.Repository)
	if err != nil {
		return nil, err
	}
	if _, err = ValidateMapping(mapping); err != nil {
		return nil, err
	}
	now := s.now().UnixMilli()
	stored := proto.Clone(mapping).(*pb.GithubStatusMapping)
	// Revision and timestamp are the store's to assign; whatever the client
	// wrote in them is discarded rather than persisted as fact.
	stored.Repository = nil
	stored.Revision = 0
	stored.UpdatedAtUnixMs = 0
	wire, err := (proto.MarshalOptions{Deterministic: true}).Marshal(stored)
	if err != nil {
		return nil, err
	}
	record, err := s.store.PutGithubStatusMapping(ctx, storage.GithubStatusMappingRecord{
		WorkspaceID: caller.WorkspaceID,
		Repository:  key(repository),
		Mapping:     wire,
		CreatedAtMS: now,
		UpdatedAtMS: now,
	}, expectedRevision)
	if err != nil {
		return nil, err
	}
	result := proto.Clone(mapping).(*pb.GithubStatusMapping)
	result.Repository = repository
	result.Revision = record.Revision
	result.UpdatedAtUnixMs = record.UpdatedAtMS
	return result, nil
}

// group finds one group by identifier.
func group(mapping *pb.GithubStatusMapping, id string) *pb.GithubStatusGroup {
	for _, candidate := range mapping.Groups {
		if candidate.Id == id {
			return candidate
		}
	}
	return nil
}

// applyLabelGroups annotates issues with the group their labels put them in.
// More than one match is reported as a conflict rather than resolved by
// picking a winner, because the configuration, not the Issue, is wrong.
func applyLabelGroups(mapping *pb.GithubStatusMapping, issues []*pb.GithubIssue) {
	if mapping.Source != pb.GithubStatusSource_GITHUB_STATUS_SOURCE_LABEL {
		return
	}
	byLabel := map[string]string{}
	for _, entry := range mapping.Groups {
		byLabel[entry.Label] = entry.Id
	}
	for _, issue := range issues {
		matched := ""
		for _, label := range issue.Labels {
			id, ok := byLabel[label.Name]
			if !ok {
				continue
			}
			if matched != "" && matched != id {
				issue.StatusConflict = true
				matched = ""
				break
			}
			matched = id
		}
		issue.StatusGroupId = matched
	}
}

// applyProjectGroups annotates issues from one Projects v2 Status field. The
// project is read once per page; an Issue that is not on the project stays
// unmapped rather than being guessed into a group.
func (s *Service) applyProjectGroups(ctx context.Context, client *githubapi.Client, mapping *pb.GithubStatusMapping, issues []*pb.GithubIssue) error {
	if mapping.Source != pb.GithubStatusSource_GITHUB_STATUS_SOURCE_PROJECT_FIELD || len(issues) == 0 {
		return nil
	}
	options, err := client.ProjectStatuses(ctx, mapping.ProjectId, mapping.ProjectFieldId)
	if err != nil {
		return s.translate(err)
	}
	byOption := map[string]string{}
	for _, entry := range mapping.Groups {
		byOption[entry.ProjectOptionId] = entry.Id
	}
	for _, issue := range issues {
		issue.StatusGroupId = byOption[options[issue.Number]]
	}
	return nil
}

File: apps/desktop/src/core/github/mapping.ts
/**
 * 状态映射：一个仓库，一个主来源，以及面板把 Issue 归在哪些分组下（设计 §7.2）。
 * 移植自 合并前的实现。
 *
 * 两个方向的耦合都可配——一个分组进入时可以设置 Issue 状态，观察到的状态也可以
 * 选出一个分组——合起来可以描述出一个环。一个环会让一次移动对着远端永远来回震荡，
 * 所以 {@link validateMapping} 在存任何东西之前先拒绝它。
 */

import {
  GithubIssueState,
  GithubStatusSource,
  type GithubIssue,
  type GithubStatusGroup,
  type GithubStatusMapping,
} from "./types";

import { githubError } from "./errors";

export const MAX_GROUPS = 32;

const GROUP_PATTERN = /^[A-Za-z0-9][A-Za-z0-9._:-]{0,63}$/;

/**
 * 完整校验一份配置。每一次拒绝都是一个**稳定的原因码**，因为这些是人要去找出来
 * 并改掉的配置错误，不是传输失败。
 *
 * 返回原因码；`""` 表示通过。不通过时同时抛 `invalid`，调用方不必两头判断。
 */
export function validateMapping(
  mapping: GithubStatusMapping | undefined,
): string {
  const reason = checkMapping(mapping);
  if (reason !== "") throw githubError("invalid");
  return reason;
}

/** 只判不抛的那一半，给「想知道原因码」的调用方和测试。 */
export function checkMapping(mapping: GithubStatusMapping | undefined): string {
  if (mapping === undefined) return "MAPPING_REQUIRED";
  if (mapping.groups.length > MAX_GROUPS) return "TOO_MANY_GROUPS";
  switch (mapping.source) {
    case GithubStatusSource.NONE:
      if (mapping.groups.length > 0 || mapping.stateGroups.length > 0) {
        return "SOURCE_NONE_HAS_GROUPS";
      }
      if (mapping.projectId !== "" || mapping.projectFieldId !== "") {
        return "SOURCE_NONE_HAS_PROJECT";
      }
      return "";
    case GithubStatusSource.LABEL:
      // 一个仓库只有一个主来源；在标签映射上带着 project 标识会让这件事含糊。
      if (mapping.projectId !== "" || mapping.projectFieldId !== "") {
        return "LABEL_SOURCE_HAS_PROJECT";
      }
      break;
    case GithubStatusSource.PROJECT_FIELD:
      if (mapping.projectId === "" || mapping.projectFieldId === "") {
        return "PROJECT_IDS_REQUIRED";
      }
      break;
    default:
      return "SOURCE_UNSPECIFIED";
  }
  if (mapping.groups.length === 0) return "GROUPS_REQUIRED";
  const ids = new Set<string>();
  const labels = new Set<string>();
  const options = new Set<string>();
  for (const group of mapping.groups) {
    if (!GROUP_PATTERN.test(group.id) || ids.has(group.id)) {
      return "GROUP_ID_INVALID";
    }
    ids.add(group.id);
    if (group.title === "" || Buffer.byteLength(group.title, "utf8") > 128) {
      return "GROUP_TITLE_INVALID";
    }
    if (
      ![
        GithubIssueState.UNSPECIFIED,
        GithubIssueState.OPEN,
        GithubIssueState.CLOSED,
      ].includes(group.couplesIssueState)
    ) {
      return "GROUP_STATE_INVALID";
    }
    if (mapping.source === GithubStatusSource.LABEL) {
      if (
        group.label === "" ||
        Buffer.byteLength(group.label, "utf8") > 128 ||
        group.projectOptionId !== ""
      ) {
        return "GROUP_LABEL_INVALID";
      }
      // 两个分组认领同一个标签会让一个 Issue 的分组**每次**都是歧义的，而不只是
      // 冲突时。
      if (labels.has(group.label)) return "GROUP_LABEL_DUPLICATE";
      labels.add(group.label);
    } else {
      if (
        group.projectOptionId === "" ||
        Buffer.byteLength(group.projectOptionId, "utf8") > 256
      ) {
        return "GROUP_OPTION_INVALID";
      }
      if (options.has(group.projectOptionId)) return "GROUP_OPTION_DUPLICATE";
      options.add(group.projectOptionId);
    }
  }
  const states = new Map<GithubIssueState, string>();
  for (const coupling of mapping.stateGroups) {
    if (
      coupling.state !== GithubIssueState.OPEN &&
      coupling.state !== GithubIssueState.CLOSED
    ) {
      return "COUPLING_STATE_INVALID";
    }
    if (!ids.has(coupling.groupId)) return "COUPLING_GROUP_UNKNOWN";
    if (states.has(coupling.state)) return "COUPLING_STATE_DUPLICATE";
    states.set(coupling.state, coupling.groupId);
  }
  return detectCycle(mapping.groups, states);
}

/**
 * 沿着「分组 → 耦合的 Issue 状态 → 那个状态的分组」走一遍。映射回自己的分组是个
 * 不动点，没问题；任何更长的环意味着进入一个分组会把 Issue 移到另一个再移回来，
 * 所以被拒。
 */
export function detectCycle(
  groups: readonly GithubStatusGroup[],
  states: ReadonlyMap<GithubIssueState, string>,
): string {
  const next = new Map<string, string>();
  for (const group of groups) {
    if (group.couplesIssueState === GithubIssueState.UNSPECIFIED) continue;
    const target = states.get(group.couplesIssueState);
    if (target === undefined || target === group.id) continue;
    next.set(group.id, target);
  }
  const UNVISITED = 0;
  const ON_PATH = 1;
  const SETTLED = 2;
  const mark = new Map<string, number>();
  for (const group of groups) {
    if ((mark.get(group.id) ?? UNVISITED) !== UNVISITED) continue;
    let node = group.id;
    const path: string[] = [];
    for (;;) {
      const state = mark.get(node) ?? UNVISITED;
      if (state === ON_PATH) return "MAPPING_CYCLE";
      if (state === SETTLED) break;
      mark.set(node, ON_PATH);
      path.push(node);
      const target = next.get(node);
      if (target === undefined) break;
      node = target;
    }
    for (const visited of path) mark.set(visited, SETTLED);
  }
  return "";
}

/** 按标识找一个分组。 */
export function findGroup(
  mapping: GithubStatusMapping,
  id: string,
): GithubStatusGroup | undefined {
  return mapping.groups.find((candidate) => candidate.id === id);
}

/**
 * 给 issue 标注它的标签把它放进了哪个分组。
 *
 * 匹配到多于一个会被报成**冲突**而不是挑一个赢家，因为错的是配置，不是这个 Issue。
 */
export function applyLabelGroups(
  mapping: GithubStatusMapping,
  issues: readonly GithubIssue[],
): void {
  if (mapping.source !== GithubStatusSource.LABEL) return;
  const byLabel = new Map<string, string>();
  for (const entry of mapping.groups) byLabel.set(entry.label, entry.id);
  for (const issue of issues) {
    let matched = "";
    for (const label of issue.labels) {
      const id = byLabel.get(label.name);
      if (id === undefined) continue;
      if (matched !== "" && matched !== id) {
        issue.statusConflict = true;
        matched = "";
        break;
      }
      matched = id;
    }
    issue.statusGroupId = matched;
  }
}

/**
 * 用一个 Projects v2 的 Status 字段标注 issue。
 *
 * project 每页只读一次；不在这个 project 上的 Issue 保持未映射，而不是被猜进一
 * 个分组。
 */
export function applyProjectGroups(
  mapping: GithubStatusMapping,
  issues: readonly GithubIssue[],
  optionsByNumber: ReadonlyMap<number, string>,
): void {
  if (mapping.source !== GithubStatusSource.PROJECT_FIELD) return;
  const byOption = new Map<string, string>();
  for (const entry of mapping.groups) {
    byOption.set(entry.projectOptionId, entry.id);
  }
  for (const issue of issues) {
    const option = optionsByNumber.get(Number(issue.number));
    issue.statusGroupId =
      option === undefined ? "" : (byOption.get(option) ?? "");
  }
}
